Thread Safety Analysis: Improved pointer handling

melver opened the pull request:
2 dependent commits:
Prior discussion: https://github.com/llvm/llvm-project/pull/123063

----------

    Thread Safety Analysis: Support warning on passing/returning pointers to guarded variables

    Introduce `-Wthread-safety-pointer` to warn when passing or returning
    pointers to guarded variables or guarded data. This is is analogous to
    `-Wthread-safety-reference`, which performs similar checks for C++
    references.

    The feature is planned to be enabled by default under `-Wthread-safety`
    in the next release cycle. This gives time for early adopters to address
    new findings.

    Adding checks for pointer passing is required to avoid false negatives
    in large C codebases, where data structures are typically implemented
    through helpers that take pointers to instances of a data structure.

---------------

    Thread Safety Analysis: Handle address-of followed by dereference

    Correctly analyze expressions where the address of a guarded variable is
    taken and immediately dereferenced, such as (*(type-specifier *)&x).
    Previously, such patterns would result in false negatives.

melver committed:
Thread Safety Analysis: Support warning on passing/returning pointers to guarded variables

Introduce `-Wthread-safety-pointer` to warn when passing or returning
pointers to guarded variables or guarded data. This is is analogous to
`-Wthread-safety-reference`, which performs similar checks for C++
references.

Adding checks for pointer passing is required to avoid false negatives
in large C codebases, where data structures are typically implemented
through helpers that take pointers to instances of a data structure.

The feature is planned to be enabled by default under `-Wthread-safety`
in the next release cycle. This gives time for early adopters to address
new findings.

Pull Request: https://github.com/llvm/llvm-project/pull/127396

melver committed:
Thread Safety Analysis: Handle address-of followed by dereference

Correctly analyze expressions where the address of a guarded variable is
taken and immediately dereferenced, such as (*(type-specifier *)&x).
Previously, such patterns would result in false negatives.

Pull Request: https://github.com/llvm/llvm-project/pull/127396

melver commented:
(Side-note: I cannot figure out what the canonical way to do "stacked commits" is for LLVM these days, so I'm just doing this the "old" way i.e. plain and simple normal git commits which I will commit separately after PR review.)

melver commented:
FWIW, I'm ready for sending a v2 series to the Linux kernel mailing list: https://git.kernel.org/pub/scm/linux/kernel/git/melver/linux.git/log/?h=cap-analysis/dev

But I'd like to sort out this PR first before sending the v2 series.

aoates commented:
As an additional data point --- I patched in these commits and have been testing with them on my C codebase and
a) so far I have no found any false positives
b) it has made a _huge_ difference in the effectiveness of TSA

I'm a strong supporter of this independent of the Linux kernel annotation work (which is, of course, also awesome).

melver commented:
Gentle ping.
Thanks!

melver commented:
I need to commit this by end of day Wednesday - if I should wait a little longer, let me know so I can plan around it.
Thanks.

AaronBallman commented:
> I need to commit this by end of day Wednesday - if I should wait a little longer, let me know so I can plan around it. Thanks.

Sorry, but I'm at C standards meetings this week, so I don't think I'll be able to review it by then. CC @aaronpuchert who maybe can help?

melver commented:
> > I need to commit this by end of day Wednesday - if I should wait a little longer, let me know so I can plan around it. Thanks.
> 
> Sorry, but I'm at C standards meetings this week, so I don't think I'll be able to review it by then. CC @aaronpuchert who maybe can help?

Thanks, and no worries. I'm also asking someone else here to help review, as we might also want to enable this in more codebases (not just kernel, which was my initial usecase).

ilya-biryukov reviewed (commented):
I only have a few comments about documenting the caveats (no alias analysis).

The actual code changes look very simple and this looks like a clear improvement that would catch many useful cases.

I don't have much experience with this code and would still advise to wait until someone more familiar with the code takes a look. But happy to give a general LGTM after the documentation is improved.

melver commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -143,6 +143,11 @@ Improvements to Clang's diagnostics
 - A statement attribute applied to a ``case`` label no longer suppresses
   'bypassing variable initialization' diagnostics (#84072).
 
+- The :doc:`ThreadSafetyAnalysis` now supports ``-Wthread-safety-pointer``
```
Note this is also still mentioned in https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#no-alias-analysis

melver commented on clang/test/SemaCXX/warn-thread-safety-analysis.cpp:
```diff
@@ -6087,9 +6215,37 @@ class Return {
+    return &foo;              // expected-warning {{returning pointer to variable 'foo' requires holding mutex 'mu' exclusively}}
+  }
+
+  Foo *returns_pt_ptr() {
```
Done.

melver commented:
> I only have a few comments about documenting the caveats (no alias analysis).
> 
> The actual code changes look very simple and this looks like a clear improvement that would catch many useful cases.
> 
> I don't have much experience with this code and would still advise to wait until someone more familiar with the code takes a look. But happy to give a general LGTM after the documentation is improved.

Added the additional test and mentioned the (remaining) lack of alias analysis.

aaronpuchert commented on clang/docs/ThreadSafetyAnalysis.rst:
```diff
@@ -515,7 +515,8 @@ Warning flags
   + ``-Wthread-safety-precise``: Requires that mutex expressions match precisely.
        This warning can be disabled for code which has a lot of aliases.
-  + ``-Wthread-safety-reference``: Checks when guarded members are passed by reference.
+  + ``-Wthread-safety-reference``: Checks when guarded members are passed or
+    returned by reference.
```
Technically, returning is covered by a subflag `-Wthread-safety-reference-return` which we apparently forget to document. On the other hand, we might also get rid of that at some point, it was mainly introduced to ease migration, not because we had evidence that users might want a warning on passing but not on returning.

aaronpuchert commented on clang/lib/Analysis/ThreadSafety.cpp:
```diff
@@ -1795,9 +1795,22 @@ void ThreadSafetyAnalyzer::checkPtAccess(const FactSet &FSet, const Expr *Exp,
 
   // Pass by reference warnings are under a different flag.
```
"reference/pointer"

aaronpuchert commented on clang/test/SemaCXX/warn-thread-safety-analysis.cpp:
```diff
@@ -4944,6 +4949,14 @@ class Foo {
+    (void)&data_ao_;      // expected-warning {{reading variable 'data_ao_' requires holding mutex 'mu_'}}
+    (void)__builtin_addressof(data_ao_); // expected-warning {{passing variable 'data_ao_' by reference requires holding mutex 'mu_'}}
+    showData(&data_ao_);  // expected-warning {{reading variable 'data_ao_' requires holding mutex 'mu_'}}
+    (void)&data_;         // no warning
+    (void)datap2_;        // no warning
+    showData(&data_);     // expected-warning {{passing pointer to variable 'data_' requires holding mutex 'mu_'}}
```
This doesn't belong here, because we're not calling overloaded operators in these cases.

aaronpuchert commented on clang/test/SemaCXX/warn-thread-safety-analysis.cpp:
```diff
@@ -4955,13 +4968,18 @@ class Foo {
 
     int a = data_[0];       // expected-warning {{reading variable 'data_' requires holding mutex 'mu_'}}
+
+    (void)&data_ao_;      // expected-warning {{reading variable 'data_ao_' requires holding mutex 'mu_'}}
+    showData(&data_ao_);  // expected-warning {{reading variable 'data_ao_' requires holding mutex 'mu_'}}
```
This is probably not so relevant. This test exists to call `const` overloads of operators, but you only defined one overload.

melver replied:
Added the second overload.

aaronpuchert commented on clang/lib/Analysis/ThreadSafety.cpp:
```diff
@@ -2294,6 +2309,11 @@ void BuildLockset::VisitReturnStmt(const ReturnStmt *S) {
         FunctionExitFSet, RetVal,
         ReturnType->getPointeeType().isConstQualified() ? AK_Read : AK_Written,
         POK_ReturnByRef);
+  } else if (ReturnType->isPointerType()) {
```
Also here, adapt the comment above to "returning by reference/pointer".

aaronpuchert reviewed (commented):
I think this looks very good! I just have some minor remarks.

Thanks to @aoates for trying this out, this is always appreciated!

And sorry for the delay.

melver commented on clang/docs/ThreadSafetyAnalysis.rst:
```diff
@@ -528,6 +529,9 @@ for a period of time, after which they are migrated into the standard analysis.
 
 * ``-Wthread-safety-beta``:  New features.  Off by default.
 
+  + ``-Wthread-safety-pointer``: Checks when passing or returning pointers to
+    guarded variables, or pointers to guarded data, as function argument or
+    return value respectively.
```
I'd wait a release cycle to turn it on by default under -Wthread-safety. This gives some time for early adopters to sort out their warnings on their own terms, rather than forcing them right now.

melver replied:
Moved the flag away from beta, but not default enabled yet.

melver commented:
> I think this looks very good! I just have some minor remarks.
> 
> Thanks to @aoates for trying this out, this is always appreciated!
> 
> And sorry for the delay.

Thanks for the review! I addressed the comments, PTAL.

Note, I think for now it might be safer to not enable by default yet, but I've made a note (and hinted at in changelog) that we're planning to default enable in future, so that should give folks enough time to deal with new findings without breaking things just yet.

Looking forward to landing this, as I'm sure it'll also uncover new findings in existing codebases using -Wthread-safety already.

aaronpuchert reviewed (approved):
Looks good to me, and thanks for the contribution!

> Note, I think for now it might be safer to not enable by default yet, but I've made a note (and hinted at in changelog) that we're planning to default enable in future, so that should give folks enough time to deal with new findings without breaking things just yet.

If it's off by default it won't be used by a lot of people, and just the separate flag should be enough to unbreak the build, should it be necessary. But I'm also fine if you'd rather leave it opt-in for a bit and collect experiences before we include it in `-Wthread-safety`.

melver commented:
Committed!

> Looks good to me, and thanks for the contribution!

Thanks for your review!
Fingers crossed the Linux kernel changes will also land soon.
 
> > Note, I think for now it might be safer to not enable by default yet, but I've made a note (and hinted at in changelog) that we're planning to default enable in future, so that should give folks enough time to deal with new findings without breaking things just yet.
> 
> If it's off by default it won't be used by a lot of people, and just the separate flag should be enough to unbreak the build, should it be necessary. But I'm also fine if you'd rather leave it opt-in for a bit and collect experiences before we include it in `-Wthread-safety`.

We'll be trying this on some internal codebases as well, and we can't yet force the cleanups to happen right away. But I expect in ~3-6 months we can flip it to default on.
